src/resources/ble_esp32.cc

```diff
@@ -1680,26 +1783,31 @@ PRIMITIVE(connect) {
 PRIMITIVE(disconnect) {
   ARGS(BleRemoteDeviceResource, device)
 
-  Locker locker(BleResourceGroup::instance()->mutex());
+  Locker locker(ble_mutex);
 
   device->disconnect();
   return process->null_object();
 }
 
 PRIMITIVE(release_resource) {
-  ARGS(Resource, resource)
+  ARGS(BleResource, resource)
 
-  Locker locker(BleResourceGroup::instance()->mutex());
+  if (resource->kind() != BleResource::ADAPTER) {
+    Locker locker(ble_mutex);
+    resource->resource_group()->unregister_resource(resource);
+  } else {
+    resource->resource_group()->unregister_resource(resource);
```

Comment:
Add a comment that explains why we can't have the locker here.

Reply:
done.